test/e2e/singlecluster/pod_test.go

```diff
@@ -108,29 +106,123 @@ var _ = ginkgo.Describe("Pod groups", func() {
-						To(gomega.BeTrueBecause("it's finished"))
-				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+
+				util.ExpectWorkloadToFinish(ctx, k8sClient, gKey)
 			})
 
 			ginkgo.By("Deleting finished Pods", func() {
-				for _, p := range g1 {
+				for _, p := range group {
 					gomega.Expect(k8sClient.Delete(ctx, p)).To(gomega.Succeed())
 				}
 				gomega.Eventually(func(g gomega.Gomega) {
-					for _, p := range g1 {
+					for _, p := range group {
 						var pCopy corev1.Pod
 						g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(p), &pCopy)).To(testing.BeNotFoundError())
 					}
 					var wl kueue.Workload
-					g.Expect(k8sClient.Get(ctx, g1Key, &wl)).Should(testing.BeNotFoundError())
+					g.Expect(k8sClient.Get(ctx, gKey, &wl)).Should(testing.BeNotFoundError())
 				}, util.Timeout, util.Interval)
 			})
 		})
+
+		ginkgo.It("Should only admit a complete group", func() {
+			group := podtesting.MakePod("group", ns.Name).
+				Image("gcr.io/k8s-staging-perf-tests/sleep:v0.1.0", []string{"1ms"}).
+				Queue(lq.Name).
+				Request(corev1.ResourceCPU, "1").
+				MakeGroup(3)
+
+			ginkgo.By("Incomplete group should not start", func() {
+				// Create incomplete group.
+				for _, p := range group[:2] {
+					gomega.Expect(k8sClient.Create(ctx, p.DeepCopy())).To(gomega.Succeed())
+				}
+				gomega.Consistently(func(g gomega.Gomega) {
+					for _, origPod := range group[:2] {
+						var p corev1.Pod
+						g.Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(origPod), &p)).To(gomega.Succeed())
+						g.Expect(p.Spec.SchedulingGates).
+							To(gomega.ContainElement(corev1.PodSchedulingGate{
+								Name: pod.SchedulingGateName}))
+					}
+				}, util.ConsistentDuration, util.Interval).Should(gomega.Succeed())
+			})
+			ginkgo.By("Incomplete group can be deleted", func() {
+				for _, p := range group[:2] {
+					gomega.Expect(k8sClient.Delete(ctx, p)).To(gomega.Succeed())
+				}
+				gomega.Eventually(func(g gomega.Gomega) {
+					for _, origPod := range group[:2] {
+						var p corev1.Pod
+						err := k8sClient.Get(ctx, client.ObjectKeyFromObject(origPod), &p)
+						g.Expect(err).To(testing.BeNotFoundError())
+					}
+				}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			})
+			ginkgo.By("Complete group runs successfully", func() {
+				for _, p := range group {
+					gomega.Expect(k8sClient.Create(ctx, p.DeepCopy())).To(gomega.Succeed())
+				}
+
+				util.ExpectWorkloadToFinish(ctx, k8sClient, client.ObjectKey{Namespace: ns.Name, Name: "group"})
+			})
+		})
+
+		ginkgo.It("Failed Pod can be replaced in group", func() {
+			group := podtesting.MakePod("group", ns.Name).
+				Image("gcr.io/k8s-staging-perf-tests/sleep:v0.1.0", []string{"1ms"}).
+				Queue(lq.Name).
+				Request(corev1.ResourceCPU, "1").
+				MakeGroup(3)
+
+			// First pod runs for much longer, so that there is time to terminate it.
+			group[0].Spec.Containers[0].Args = []string{"-termination-code=1", "10m"}
```

Comment:
I confirmed manually that this produces a Pod with exit code 1 when calling Delete.